File: js/tokenmaster/src/events.ts
/**
 * Typed event stream, per docs/core-api.md section 4.
 *
 * Wire shape for every event:
 *
 *     {"event_type": ..., "schema_version": ..., "timestamp": ...,
 *      "turn_id": ..., "payload": {...}}
 *
 * This stream is the entire contract between tokenmaster and any visualizer.
 * Every event type something can emit today is implemented here; only
 * CalibrationLoaded awaits its feature, so that no event type exists in code
 * before something emits it.
 *
 * Timestamps are ISO 8601 via Date.toISOString() ("Z" suffix); the Python
 * reference emits "+00:00". Conformance is unaffected: the spec excludes
 * timestamps from comparison everywhere, and wire round trips preserve the
 * stored string verbatim.
 */

import {
  SCHEMA_VERSION,
  MeterState,
  TurnUsage,
  Zone,
  asZone,
} from "./types.js";
import { Recommendation } from "./advisor.js";
import { FidelityReport } from "./fidelity.js";

function utcnow(): string {
  return new Date().toISOString();
}

function asFloat(value: unknown, field: string): number {
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n)) {
    throw new TypeError(`${field} must be a finite number`);
  }
  return n;
}

function reqString(value: unknown, field: string): string {
  if (value === null || value === undefined) {
    throw new TypeError(`${field} is required`);
  }
  return String(value);
}

// ---------------------------------------------------------------------------
// envelope

export interface EventDict {
  event_type: string;
  schema_version: string;
  timestamp: string;
  turn_id: number | null;
  payload: Record<string, unknown>;
}

/** Envelope fields shared by every event constructor. */
export interface EventInit {
  turn_id?: number | null;
  timestamp?: string;
  schema_version?: string;
}

export type EventCallback = (event: Event) => void;

/** Base envelope. Subclasses define EVENT_TYPE and payload fields. */
export abstract class Event {
  static readonly EVENT_TYPE: string = "event";

  readonly turn_id: number | null;
  readonly timestamp: string;
  readonly schema_version: string;

  constructor(fields: EventInit = {}) {
    this.turn_id = fields.turn_id ?? null;
    this.timestamp = fields.timestamp ?? utcnow();
    this.schema_version = fields.schema_version ?? SCHEMA_VERSION;
  }

  get event_type(): string {
    return (this.constructor as typeof Event).EVENT_TYPE;
  }

  payload(): Record<string, unknown> {
    return {};
  }

  toDict(): EventDict {
    return {
      event_type: this.event_type,
      schema_version: this.schema_version,
      timestamp: this.timestamp,
      turn_id: this.turn_id,
      payload: this.payload(),
    };
  }

  toJSON(): EventDict {
    return this.toDict();
  }
}

// ---------------------------------------------------------------------------
// meter-emitted events

/** A turn was ingested; carries the turn and the resulting state. */
export class TurnRecorded extends Event {
  static readonly EVENT_TYPE = "turn_recorded";

  readonly turn: TurnUsage;
  readonly state: MeterState;

  constructor(fields: EventInit & { turn: TurnUsage; state: MeterState }) {
    super(fields);
    this.turn = fields.turn;
    this.state = fields.state;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return { turn: this.turn.toDict(), state: this.state.toDict() };
  }
}

/** fill_effective crossed a zone boundary. */
export class ZoneChanged extends Event {
  static readonly EVENT_TYPE = "zone_changed";

  readonly from_zone: Zone;
  readonly to_zone: Zone;
  readonly fill_effective: number;

  constructor(
    fields: EventInit & {
      from_zone: Zone;
      to_zone: Zone;
      fill_effective: number;
    }
  ) {
    super(fields);
    this.from_zone = fields.from_zone;
    this.to_zone = fields.to_zone;
    this.fill_effective = fields.fill_effective;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return {
      from_zone: this.from_zone,
      to_zone: this.to_zone,
      fill_effective: this.fill_effective,
    };
  }
}

/** Velocity moved by more than the configured factor between turns. */
export class VelocityShift extends Event {
  static readonly EVENT_TYPE = "velocity_shift";

  readonly previous: number;
  readonly current: number;

  constructor(fields: EventInit & { previous: number; current: number }) {
    super(fields);
    this.previous = fields.previous;
    this.current = fields.current;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return { previous: this.previous, current: this.current };
  }
}

/**
 * A recorded turn carried a different model_id than the previous one.
 *
 * The Meter keeps gauging against its constructed profile; this event only
 * reports the switch so consumers can decide what it means for them.
 */
export class ModelChanged extends Event {
  static readonly EVENT_TYPE = "model_changed";

  readonly previous_model_id: string;
  readonly new_model_id: string;

  constructor(
    fields: EventInit & { previous_model_id: string; new_model_id: string }
  ) {
    super(fields);
    this.previous_model_id = fields.previous_model_id;
    this.new_model_id = fields.new_model_id;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return {
      previous_model_id: this.previous_model_id,
      new_model_id: this.new_model_id,
    };
  }
}

// ---------------------------------------------------------------------------
// advisor events

/** A policy was evaluated; carries the full recommendation. */
export class AdvisorRecommendation extends Event {
  static readonly EVENT_TYPE = "advisor_recommendation";

  readonly recommendation: Recommendation;

  constructor(fields: EventInit & { recommendation: Recommendation }) {
    super(fields);
    this.recommendation = fields.recommendation;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return { recommendation: this.recommendation.toDict() };
  }
}

// ---------------------------------------------------------------------------
// fidelity events

/** A handoff artifact was scored; carries the full fidelity report. */
export class HandoffEvaluated extends Event {
  static readonly EVENT_TYPE = "handoff_evaluated";

  readonly report: FidelityReport;

  constructor(fields: EventInit & { report: FidelityReport }) {
    super(fields);
    this.report = fields.report;
    Object.freeze(this);
  }

  payload(): Record<string, unknown> {
    return { report: this.report.toDict() };
  }
}

// ---------------------------------------------------------------------------
// wire reconstruction

type EnvelopeFields = Required<EventInit>;

const EVENT_FACTORIES: Record<
  string,
  (envelope: EnvelopeFields, payload: Record<string, unknown>) => Event
> = {
  [TurnRecorded.EVENT_TYPE]: (envelope, payload) =>
    new TurnRecorded({
      ...envelope,
      turn: TurnUsage.fromDict(payload["turn"] as Record<string, unknown>),
      state: MeterState.fromDict(payload["state"] as Record<string, unknown>),
    }),
  [ZoneChanged.EVENT_TYPE]: (envelope, payload) =>
    new ZoneChanged({
      ...envelope,
      from_zone: asZone(payload["from_zone"]),
      to_zone: asZone(payload["to_zone"]),
      fill_effective: asFloat(payload["fill_effective"], "fill_effective"),
    }),
  [VelocityShift.EVENT_TYPE]: (envelope, payload) =>
    new VelocityShift({
      ...envelope,
      previous: asFloat(payload["previous"], "previous"),
      current: asFloat(payload["current"], "current"),
    }),
  [ModelChanged.EVENT_TYPE]: (envelope, payload) =>
    new ModelChanged({
      ...envelope,
      previous_model_id: reqString(
        payload["previous_model_id"],
        "previous_model_id"
      ),
      new_model_id: reqString(payload["new_model_id"], "new_model_id"),
    }),
  [AdvisorRecommendation.EVENT_TYPE]: (envelope, payload) =>
    new AdvisorRecommendation({
      ...envelope,
      recommendation: Recommendation.fromDict(
        (payload["recommendation"] as Record<string, unknown>) ?? {}
      ),
    }),
  [HandoffEvaluated.EVENT_TYPE]: (envelope, payload) =>
    new HandoffEvaluated({
      ...envelope,
      report: FidelityReport.fromDict(
        (payload["report"] as Record<string, unknown>) ?? {}
      ),
    }),
};

/** Reconstruct a typed event from its wire dictionary. */
export function eventFromDict(dict: object): Event {
  const d = dict as Record<string, unknown>;
  const eventType = d["event_type"];
  const factory =
    typeof eventType === "string" ? EVENT_FACTORIES[eventType] : undefined;
  if (factory === undefined) {
    throw new RangeError(`Unknown event_type: '${String(eventType)}'`);
  }
  return factory(
    {
      turn_id: (d["turn_id"] ?? null) as number | null,
      timestamp: reqString(d["timestamp"], "timestamp"),
      schema_version:
        d["schema_version"] === null || d["schema_version"] === undefined
          ? SCHEMA_VERSION
          : String(d["schema_version"]),
    },
    (d["payload"] ?? {}) as Record<string, unknown>
  );
}

File: rust/tokenmaster/src/events.rs
//! Typed event stream, per docs/core-api.md section 4.
//!
//! Wire shape for every event:
//!
//! ```text
//! {"event_type": ..., "schema_version": ..., "timestamp": ...,
//!  "turn_id": ..., "payload": {...}}
//! ```
//!
//! This stream is the entire contract between tokenmaster and any
//! visualizer. Events implemented here are the ones the Meter can emit at
//! this point in the port (TurnRecorded, ZoneChanged, VelocityShift,
//! ModelChanged); AdvisorRecommendation and HandoffEvaluated arrive with the
//! advisor and fidelity modules, and CalibrationLoaded awaits its feature,
//! so that no event type exists in code before something emits it.
//!
//! Timestamps are ISO 8601 UTC with a "Z" suffix and microsecond precision,
//! produced by a std-only epoch-to-civil conversion (the reference emits
//! "+00:00" via datetime, the JS port "Z" via Date). Conformance is
//! unaffected: the spec excludes timestamps from comparison everywhere, and
//! wire round trips preserve the stored string verbatim.

use std::str::FromStr;
use std::time::{SystemTime, UNIX_EPOCH};

use serde::Serialize;
use serde_json::{Map, Value};

use crate::types::{
    as_map, opt_i64, req_f64, req_string, string_or, Error, MeterState, TurnUsage, Zone,
    SCHEMA_VERSION,
};

// ------------------------------------------------------------------------ //
// timestamps

/// Gregorian civil date from days since 1970-01-01 (Howard Hinnant's
/// civil_from_days).
fn civil_from_days(days: i64) -> (i64, u32, u32) {
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1_460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 { mp + 3 } else { mp - 9 } as u32;
    (if m <= 2 { y + 1 } else { y }, m, d)
}

fn format_utc(secs: i64, micros: u32) -> String {
    let days = secs.div_euclid(86_400);
    let rem = secs.rem_euclid(86_400);
    let (year, month, day) = civil_from_days(days);
    let (h, m, s) = (rem / 3_600, (rem % 3_600) / 60, rem % 60);
    format!("{year:04}-{month:02}-{day:02}T{h:02}:{m:02}:{s:02}.{micros:06}Z")
}

/// Current UTC time as an ISO 8601 string.
pub(crate) fn utcnow() -> String {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    format_utc(now.as_secs() as i64, now.subsec_micros())
}

// ------------------------------------------------------------------------ //
// events

/// One emitted event: the shared envelope plus a typed payload.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Event {
    pub turn_id: Option<i64>,
    pub timestamp: String,
    pub schema_version: String,
    #[serde(flatten)]
    pub kind: EventKind,
}

/// Typed payloads. Wire tag is `event_type`, payload fields nest under
/// `payload`.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "event_type", content = "payload", rename_all = "snake_case")]
pub enum EventKind {
    /// A turn was ingested; carries the turn and the resulting state.
    TurnRecorded { turn: TurnUsage, state: MeterState },
    /// fill_effective crossed a zone boundary.
    ZoneChanged {
        from_zone: Zone,
        to_zone: Zone,
        fill_effective: f64,
    },
    /// Velocity moved by more than the configured factor between turns.
    VelocityShift { previous: f64, current: f64 },
    /// A recorded turn carried a different model_id than the previous one.
    ///
    /// The Meter keeps gauging against its constructed profile; this event
    /// only reports the switch so consumers can decide what it means for
    /// them.
    ModelChanged {
        previous_model_id: String,
        new_model_id: String,
    },
}

impl EventKind {
    /// Wire tag for this payload.
    pub fn event_type(&self) -> &'static str {
        match self {
            EventKind::TurnRecorded { .. } => "turn_recorded",
            EventKind::ZoneChanged { .. } => "zone_changed",
            EventKind::VelocityShift { .. } => "velocity_shift",
            EventKind::ModelChanged { .. } => "model_changed",
        }
    }
}

fn req_value<'a>(d: &'a Map<String, Value>, key: &str, ctx: &str) -> Result<&'a Value, Error> {
    d.get(key)
        .ok_or_else(|| Error::Parse(format!("{ctx}: missing required field '{key}'")))
}

impl Event {
    /// Construct an event stamped with the current time.
    pub fn new(turn_id: Option<i64>, kind: EventKind) -> Event {
        Event {
            turn_id,
            timestamp: utcnow(),
            schema_version: SCHEMA_VERSION.to_string(),
            kind,
        }
    }

    /// Wire tag for this event.
    pub fn event_type(&self) -> &'static str {
        self.kind.event_type()
    }

    /// JSON string form. Infallible for any event the core produces.
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("Event serialization")
    }

    /// Reconstruct a typed event from its wire form (the reference's
    /// `event_from_dict`).
    pub fn from_value(v: &Value) -> Result<Event, Error> {
        let d = as_map(v, "Event")?;
        let event_type = req_string(d, "event_type", "Event")?;
        let empty = Map::new();
        let payload = match d.get("payload") {
            None => &empty,
            Some(p) => as_map(p, "Event payload")?,
        };
        let kind = match event_type.as_str() {
            "turn_recorded" => EventKind::TurnRecorded {
                turn: TurnUsage::from_value(req_value(payload, "turn", "TurnRecorded payload")?)?,
                state: MeterState::from_value(req_value(
                    payload,
                    "state",
                    "TurnRecorded payload",
                )?)?,
            },
            "zone_changed" => EventKind::ZoneChanged {
                from_zone: Zone::from_str(&req_string(
                    payload,
                    "from_zone",
                    "ZoneChanged payload",
                )?)?,
                to_zone: Zone::from_str(&req_string(payload, "to_zone", "ZoneChanged payload")?)?,
                fill_effective: req_f64(payload, "fill_effective", "ZoneChanged payload")?,
            },
            "velocity_shift" => EventKind::VelocityShift {
                previous: req_f64(payload, "previous", "VelocityShift payload")?,
                current: req_f64(payload, "current", "VelocityShift payload")?,
            },
            "model_changed" => EventKind::ModelChanged {
                previous_model_id: req_string(
                    payload,
                    "previous_model_id",
                    "ModelChanged payload",
                )?,
                new_model_id: req_string(payload, "new_model_id", "ModelChanged payload")?,
            },
            other => return Err(Error::Value(format!("Unknown event_type: '{other}'"))),
        };
        Ok(Event {
            turn_id: opt_i64(d, "turn_id", "Event")?,
            timestamp: req_string(d, "timestamp", "Event")?,
            schema_version: string_or(d, "schema_version", SCHEMA_VERSION, "Event")?,
            kind,
        })
    }
}

// ------------------------------------------------------------------------ //
// unit tests for the timestamp conversion (pinned known epochs)

#[cfg(test)]
mod tests {
    use super::{format_utc, utcnow};

    #[test]
    fn format_utc_matches_known_epochs() {
        assert_eq!(format_utc(0, 0), "1970-01-01T00:00:00.000000Z");
        assert_eq!(format_utc(1_767_225_600, 0), "2026-01-01T00:00:00.000000Z");
        // Leap day, midday, with microseconds.
        assert_eq!(
            format_utc(1_709_164_800 + 43_200, 123_456),
            "2024-02-29T12:00:00.123456Z"
        );
    }

    #[test]
    fn utcnow_is_iso_8601_utc() {
        let ts = utcnow();
        assert_eq!(ts.len(), "1970-01-01T00:00:00.000000Z".len());
        assert!(ts.ends_with('Z'));
        assert_eq!(&ts[10..11], "T");
    }
}
